Bug fixes in Solovay Kitaev Decomposition

alexanderivrii opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

This PR fixes a number of bugs we have recently discovered in the Python code for the Solovay-Kitaev decomposition, together with @Cryoris and @anedumla.

Fixes include:
* fixed a problem with the rotation axis computation for 180 degree rotations (thanks to @Cryoris for working out the math)
* fixed a problem where the global phase of the returned circuit might have been off by $\pi$, due to the underlying algorithm approximating not the original ``SU(2)`` matrix, but its projection onto ``SO(3)``
* in particular fixed #9552.

Longer-term, we will deprecate the Python code for Solovay-Kitaev decomposition altogether, and switch to the new improved pass written in Rust that also fixes these problems (see #14203), however for backward compatibility we need to temporarily support the Python code as well, and thus it makes sense to fix the problems on the Python side too.


### Details and comments

Other changes:

* There was a very strange "back door" for returning a circuit with the original unitary matrix if the computed matrix is not close to the identity. This behavior is not consistent with the documented description of the pass and was removed. (Or rather the functions ``to_circuit`` and ``to_dag`` remain for backwards compatibility but are no longer used by the main pass).
* The ``GateSequence`` class used to store both the SO(3) matrix as ``product`` and the SU(2) matrix as ``product_su2``, however the parameter ``product_su2`` was not updated correctly in various places (and it was not even clear what to do within ``GateSequence.from_matrix``). As ``product_su2`` was almost not used anywhere in the algorithm (with the exception of one place where its usage did not seem to be correct anyway), it was simply removed to avoid confusion and unnecessary computations.

alexanderivrii committed:
Fix _compute_rotation_axis method

The method used to return [0, 0, 0] for rotation axis when theta is very close to 0 (and yet not within 1e010).
This commit fixes this by noting that the rotation axis of an SO(3) matrix is simply the eigenvector of this
matrix corresponding to eigenvalue 1.

alexanderivrii committed:
Fixes related to SU(2) vs. SO(3)

The implemented algorithm has a global phase uncertainty of +-1, due to approximating not the
original SU(2) matrix but its projection onto SO(3). This fixes the global phase of the
computed approximation.

In addition, the product_su2 matrix in GateSequence was not correctly computed in many cases,
leading to incorrect values when using this attribute. Since this is not needed for the actual
algorithm (and spends unnecessary effort for computing it), this also completely removes this
attribute.

Co-authored-by: Almudena Carrera Vazquez <almudenacarreravazquez@hotmail.com>

alexanderivrii committed:
Fix the tests.

One of the tests was simply wrong because the reference circuit has the wrong global phase.
The tests that check the gates in the final decomposition should not assume that SGate is included and SdgGate is not.

alexanderivrii committed:
reno

jakelishman commented:
This isn't a review, but if the purpose of this is a bugfix to touch the Python code that we're removing in favour of Rust code, should it be targetting as API stable for backport to 1.4 and 2.0?

alexanderivrii committed:
fix to the global phase

alexanderivrii committed:
bug fix (forgetting that gate_matrix_su2 is not SU(2) was a sequence)

alexanderivrii committed:
adding test

alexanderivrii commented:
@jakelishman, yes, I believe we should. Does this mean that we should change the milestone to 2.0? Or keep it as 2.1 and talk to mergify?

alexanderivrii commented:
One more problem that I have recently discovered (and fixed in the followup PR #14225) is that the `SolovayKitaevSynthesis` unitary synthesis plugin caches the computed approximations regardless of the basis set, that is the following code:
```
circuit = QuantumCircuit(1)
circuit.rx(0.8, 0)
unitary = Operator(circuit).data
plugin = SolovayKitaevSynthesis()

# First, use the ["h", "t", "tdg"] set
out = plugin.run(unitary, basis_gates=["h", "t", "tdg"])
print(out.count_ops())
self.assertLessEqual(set(out.count_ops().keys()), {"h", "t", "tdg"})

# Second, use the ["h", "s"] set
out = plugin.run(unitary, basis_gates=["h", "s", "sdg"])
self.assertLessEqual(set(out.count_ops().keys()), {"h", "s", "sdg"})
print(out.count_ops())
```
synthesizes both times to the first specified basis set of ``["h", "t", "tdg"]``. Creating two different ``SolovayKitaevSynthesis()`` plugins does not solve this, since the caching is done at a class-level attribute. The fix (many thanks to @eliarbel for the suggestion!) consists of storing a dictionary from the basis sets to cached approximations instead. I am wondering if it would make sense to include this fix in this PR instead?

Cryoris commented on qiskit/synthesis/discrete_basis/gate_sequence.py:
```diff
@@ -118,6 +142,10 @@ def to_dag(self):
         dag = DAGCircuit()
         dag.add_qubits(qreg)
 
+        # Note: returning the circuit with the unitary matrix instead of its decomposition
```
I don't understand this comment: If the sequence is initialized with a matrix and does not have any gates in it, then we can only convert it into a circuit/dag by using a `UnitaryGate`, no?

alexanderivrii replied:
Personally, I did not like the fact that `GateSequence` serves two different purposes. First, as documented in the class docstring, "the class stores the sequence of gates along with the unitary they implement". This is the behavior that I expect: the attribute `gates` stores a list of gates in the sequence, and the attribute `product` stores the SO(3)-product of the corresponding matrices.

However, we also use `GateSequence` to internally store the single SO(3)-matrix at the start of `SolovayKitaevDecomposition.run`. In this case, `gates` is an empty list and `product` is the matrix we want to store. Now the statement "the class stores the sequence of gates along with the unitary they implement" is no longer true. I think that instead of (ab)using the `GateSequence` in this way, we should change the argument to `SolovayKitaev.recurse` to accept a matrix and not a gate sequence (but I have not thoroughly thought through this approach).

In any case, the decomposition should **not** return circuits with unitary matrices -- the returned circuits should only contain the gates passed in the approximations, do you agree?. Furthermore, in the case of basis gates such as S/H, we may very well have the case that the empty gate list is the closest approximation (yet not sufficiently close to identity), again we do not want to produce a unitary instead. 

Ideally, I would suggest to remove the ``if len(self.gates) == 0 and not np.allclose(self.product, np.identity(3)):`` from `to_circuit` and `to_dag`, but this could be considered a breaking change.

alexanderivrii replied:
Experimenting with the above, I can locally make the code work after the following simplifications:
* remove `GateSequence::from_matrix` method
* change the function `commutator_decompose` to return a pair of SO(3)-metrices `v`, `w` instead of `GateSequence` objects
* Change the inputs to `SKD::_recurse` and `SKD::find_basic_approximations` to be SO(3)-matrices; keep the outputs as `GateSequence`
* A few more minor changes of using matrices instead of sequences

I think this makes the code much cleaner and easier to understand, but unfortunately some of these changes would break the documented API. What do you think?

Cryoris replied:
I would try to reduce only to bugfixes here (remember that the main purpose of this PR is for backporting, as for the 2.x series we'll have the Rust version of SK), so I wouldn't change the behavior here. It's true that `GateSequence` serves two purposes here, though I don't think it's too bad -- it could maybe be improved by setting `gates` to `None` if we don't have a gate sequence yet (but again not for this PR 🙂).

I'm not sure I understand the S/H case: if the identity is the best approximation and gates is empty, then it should return the empty circuit, not the unitary, since we have the `not np.allclose(product, I)` bit, no?

Cryoris replied:
Adding to the above: then we also don't need `_to_dag` and `_to_circuit` and can just keep the existing `to_dag` and `to_circuit` 🙂

Cryoris commented on qiskit/synthesis/discrete_basis/gate_sequence.py:
```diff
@@ -129,6 +157,23 @@ def to_dag(self):
 
         return dag
 
+    def _to_dag(self, adjust_phase: bool):
```
See comment above; maybe this is not needed and we can just use the above `to_dag` method (with the `adjust_phase` fix).

alexanderivrii replied:
So now I could delete the new functions `_to_circuit` and `_to_dag` but see the reply to the comment above (after I write it).

Cryoris commented on qiskit/synthesis/discrete_basis/solovay_kitaev.py:
```diff
@@ -155,17 +159,20 @@ def _recurse(self, sequence: GateSequence, n: int, check_input: bool = True) ->
 
         if n == 0:
-            return self.find_basic_approximation(sequence)
+            res = self.find_basic_approximation(sequence)
```
IMO the `return` statement was a bit cleaner, but I don't have a very strong opinion 🙂

alexanderivrii replied:
Technically I did not need to change that, but for some reason I personally prefer this a bit more (but also I do not have a strong opinion).

Cryoris commented on qiskit/synthesis/discrete_basis/gate_sequence.py:
```diff
@@ -106,6 +109,27 @@ def to_circuit(self):
+        u2 = np.exp(1j * self.global_phase) * u2
+        return u2
+
+    def _to_circuit(self, adjust_phase: bool):
```
Could we just move the phase fixing inside this function (and the `_to_dag`)? Or do you think there's a scenario where one would not like to adjust the phase?

alexanderivrii replied:
That's fair. In [683f127](https://github.com/Qiskit/qiskit/pull/14217/commits/683f127592208cbc220ea86d302c778b80e34a8d) I have improved the implementation to do all the phase fixing in `SolovayKitaevDecomposition.run`, including updating the SO(3)-phase to the correct SU(2)-phase, and adjusting that to the original U(2)-phase.

Cryoris commented on qiskit/synthesis/discrete_basis/commutator_decompose.py:
```diff
@@ -53,17 +53,15 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
-        y = 0.0
-        z = 0.0
-    return np.array([x, y, z])
+    # A simple (but probably not the most efficient) way to compute the rotation
```
Regarding our offline convo; if you agree then using the Rodrigues path is probably faster 🙂

alexanderivrii replied:
I agree. In [a8a1954](https://github.com/Qiskit/qiskit/pull/14217/commits/a8a1954c097546f24f80ba106c87c7f57e9ac533) I am restoring the previous code while also handling the 180-degree rotation based on your computation (thanks!).

Cryoris reviewed (commented):
Thanks for taking care of these! And I agree these should be backported, especially now that we decided we'll do the oxidization in place.

alexanderivrii committed:
Restoring the previous fast code for _compute_rotation_axis based on Rodrigues formula, while additionally handling the case of 180-degree rotation

alexanderivrii committed:
removing old code

alexanderivrii committed:
review comments

Cryoris commented on qiskit/synthesis/discrete_basis/commutator_decompose.py:
```diff
@@ -53,16 +53,42 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
+        # The matrix is the 180-degree rotation
+        # We have:
+        #   k_x^2 = (1 + r_00) / 2, k_y^2 = (1 + r_11) / 2, k_z^2 = (1 + r_22) / 2,
+        #   k_x k_y = r_01 / 2, k_y k_z = r_12 / 2, k_x k_z = r_02 / 2
```
The formula's are a bit out of context here, I'd either remove them or explain a bit more what `k` and `r` are 🙂

alexanderivrii replied:
Ok, I will remove it.

Cryoris commented on qiskit/synthesis/discrete_basis/commutator_decompose.py:
```diff
@@ -53,16 +53,42 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
+        # We have:
+        #   k_x^2 = (1 + r_00) / 2, k_y^2 = (1 + r_11) / 2, k_z^2 = (1 + r_22) / 2,
+        #   k_x k_y = r_01 / 2, k_y k_z = r_12 / 2, k_x k_z = r_02 / 2
+        squares = [0.5 * (1 + matrix[0][0]), 0.5 * (1 + matrix[1][1]), 0.5 * (1 + matrix[2][2])]
```
```suggestion
        squares = (1 + np.diagonal(matrix)) / 2
```

alexanderivrii replied:
this looks nicer indeed

Cryoris commented on qiskit/synthesis/discrete_basis/commutator_decompose.py:
```diff
@@ -53,16 +53,42 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
+        #   k_x^2 = (1 + r_00) / 2, k_y^2 = (1 + r_11) / 2, k_z^2 = (1 + r_22) / 2,
+        #   k_x k_y = r_01 / 2, k_y k_z = r_12 / 2, k_x k_z = r_02 / 2
+        squares = [0.5 * (1 + matrix[0][0]), 0.5 * (1 + matrix[1][1]), 0.5 * (1 + matrix[2][2])]
+        index_of_max = squares.index(max(squares))
```
There's a function for this 🙂 
```suggestion
        index_of_max = np.argmax(squares)
```

alexanderivrii replied:
nice to know, this is nicer indeed

Cryoris commented on qiskit/synthesis/discrete_basis/commutator_decompose.py:
```diff
@@ -53,16 +53,42 @@ def _compute_rotation_axis(matrix: np.ndarray) -> np.ndarray:
+    else:
+        # The matrix is the rotation by theta with sin(theta)!=0
+        theta = math.acos(0.5 * (trace - 1))
+        x = 1 / (2 * math.sin(theta)) * (matrix[2][1] - matrix[1][2])
```
We should use the antisymmetrized matrix here, right? I.e. `r = (matrix - matrix.T) / 2` instead of `matrix`

alexanderivrii replied:
This is how it was before my changes, so (your argument on minimizing changes) let's not change it. But yes, ``matrix[2][1] - matrix[1][2]`` is exactly an entry in ``matrix - matrix.T``. 

(Note: as the sign is important, the entry to choose is the entry ``matrix[2][1] - matrix[1][2]`` and not ``matrix[1][2] - matrix[2][1]``, as follows from the calculation based on Rodrigues rotation formula).

Cryoris commented on qiskit/synthesis/discrete_basis/generate_basis_approximations.py:
```diff
@@ -76,7 +76,7 @@ def _check_candidate_greedy(candidate, existing_sequences, tol=1e-10):
 
     for existing in existing_sequences:
-        if matrix_equal(existing.product_su2, candidate.product_su2, ignore_phase=True, atol=tol):
+        if matrix_equal(existing.product, candidate.product, ignore_phase=False, atol=tol):
```
I'm wondering if `ignore_phase` even matters here, since the SO(3) product does not have a phase. In this case I'd rather not introduce a change

alexanderivrii replied:
Agreed, changed it back.

alexanderivrii committed:
switching back to using _to_dag and _to_circuit

I actually no longer think that these methods would ever return a circuit with unitary gates, so it does not really matter

alexanderivrii committed:
Addressing review comments

alexanderivrii commented:
For the purposes of this PR, I changed back to using the existing `to_dag` and `to_circuit`. Indeed, in the discussed S/H-case we just get back the empty circuit, which is the correct answer (for some reason I thought that this is not what would happen). I don't think anymore that the main run method of Solovay-Kitaev can return a circuit with a unitary gate, so this change does not even matter.

For the record: I do think that it would be best to simplify the arguments to different functions (those that need to take matrices should take matrices and not gate sequences), however I agree with you that there is no point to work on this in the Python space, as we are in the process of replacing this code by a much better Rust version.

Cryoris commented on releasenotes/notes/fix-sk-decomposition-23da3ee4b6a10d62.yaml:
```diff
@@ -0,0 +1,8 @@
+fixes:
+  - |
+    Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the pass would crash due
+    to incorrectly computing the unitary matrix corresponding to a 180 degree rotation.
```
```suggestion
    Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the pass could
    randomly crash due encountering a 180 degree rotation in the internal recursion,
    which was not handled correctly. 
```

alexanderivrii replied:
improved in [056e19a](https://github.com/Qiskit/qiskit/pull/14217/commits/056e19ac020fc01e13a75733ae6f999a91aa708d)

Cryoris commented on releasenotes/notes/fix-sk-decomposition-23da3ee4b6a10d62.yaml:
```diff
@@ -0,0 +1,8 @@
+    to incorrectly computing the unitary matrix corresponding to a 180 degree rotation.
+  - Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the pass would produce
+    an incorrect global phase due to internally approximating the SO(3) and not the SU(2) matrices.
+    Fixed `#9552 <https://github.com/Qiskit/qiskit-terra/issues/9552>`__
```
```suggestion
  - Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the pass could sometimes
    produce an approximation where the phase is off by :math:`\pi`. This is due to the internal 
    :math:`SO(3)` representation, which requires special care to obtain the correct sign of the 
    qubit gate matrix.
    Fixed `#9552 <https://github.com/Qiskit/qiskit-terra/issues/9552>`__
```

alexanderivrii replied:
improved in [056e19a](https://github.com/Qiskit/qiskit/pull/14217/commits/056e19ac020fc01e13a75733ae6f999a91aa708d)

alexanderivrii committed:
pass over release notes combining the original and the suggested wordings

Cryoris commented on releasenotes/notes/fix-sk-decomposition-23da3ee4b6a10d62.yaml:
```diff
@@ -0,0 +1,12 @@
+  - |
+    Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the generated
+    approximation could have a phase that differs by :math:`\pi` from the correct value.
+    This resulted due the internal :math:`SO(3)` representation, which requires additional
```
Two minor typos (I'm not committing since I'm not co-author) 🙂 
```suggestion
    crash due to encountering a 180 degree rotation in the internal recursion,
    which was not handled correctly.
  - |
    Fixed a problem in the :class:`.SolovayKitaev` transpiler pass where the generated
    approximation could have a phase that differs by :math:`\pi` from the correct value.
    This resulted due to the internal :math:`SO(3)` representation, which requires additional
```

alexanderivrii committed:
Update releasenotes/notes/fix-sk-decomposition-23da3ee4b6a10d62.yaml

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris reviewed (approved):
LGTM thanks for integrating the fixes!

mtreinish commented:
@Mergifyio backport stable/1.4
